File: 0733. Flood Fill/0733.java
// time complexity: O(m x n)
// space complexity: O(m x n)
class Solution {
    // DFS RECURSION ALGORITHM
    public int[][] floodFill(int[][] image, int sr, int sc, int color) {
        int oldColor = image[sr][sc];
        // if the starting pixel is already the target color, no work needed
        if (oldColor == color) return image;
        dfs(image, sr, sc, oldColor, color);
        return image;
    }

    private void dfs(int[][] img, int r, int c, int oldColor, int newColor) {
        // boundary check and color match check
        if (r < 0 || r >= img.length || c < 0 || c >= img[0].length) return;
        if (img[r][c] != oldColor) return;

        img[r][c] = newColor;          // fill current pixel
        // recurse on 4-connected neighbors
        dfs(img, r + 1, c, oldColor, newColor);
        dfs(img, r - 1, c, oldColor, newColor);
        dfs(img, r, c + 1, oldColor, newColor);
        dfs(img, r, c - 1, oldColor, newColor);
    }

}

File: 0733. Flood Fill/0733.py
# time complexity: O(m x n)
# space complexity: O(m x n)
class Solution(object):
    # BFS ALGORITHM with a QUEUE
    def floodFill(self, image, start_row, start_col, new_color):
        """
        :type image: List[List[int]]
        :type sr: int
        :type sc: int
        :type color: int
        :rtype: List[List[int]]
        """
        original_color = image[start_row][start_col]
    
        # If the starting pixel already has the target color, nothing to do
        if original_color == new_color:
            return image

        rows, cols = len(image), len(image[0])
        # Queue holds pixels to process; start with the initial pixel
        queue = deque()
        queue.append((start_row, start_col))
        # Color the start pixel immediately
        image[start_row][start_col] = new_color

        # Directions for 4-connected neighbors: down, up, right, left
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]

        # Continue until no more pixels in queue
        while queue:
            current_row, current_col = queue.popleft()
            
            # Check all 4 neighbors
            for dr, dc in directions:
                neighbor_row = current_row + dr
                neighbor_col = current_col + dc
                
                # Check bounds
                if (0 <= neighbor_row < rows) and (0 <= neighbor_col < cols):
                    # If the neighbor has the original color, fill and enqueue it
                    if image[neighbor_row][neighbor_col] == original_color:
                        image[neighbor_row][neighbor_col] = new_color
                        queue.append((neighbor_row, neighbor_col))

        return image